### ISSUE TITLE
Teuchos parser unit test compile fails with gcc 5.4 and CUDA

### ISSUE THREAD

**Issue by jjellio** [trilinos/Trilinos#1897]
Can reproduce on white:

module load devpack-sparc/openmpi/1.10.4/gcc/5.4.0/cuda/8.0.4

A suitable build script is attached. (source it)

```
[ 75%] Building CXX object packages/teuchos/parser/test/CMakeFiles/TeuchosParser_Parser_UnitTests.dir/unit_tests.cpp.o
/ascldap/users/jjellio/src/Trilinos/packages/teuchos/parser/test/unit_tests.cpp(146): error: "<unnamed>::any" is ambiguous

/ascldap/users/jjellio/src/Trilinos/packages/teuchos/parser/test/unit_tests.cpp(226): error: "<unnamed>::any" is ambiguous

/ascldap/users/jjellio/src/Trilinos/packages/teuchos/parser/test/unit_tests.cpp(240): error: "<unnamed>::any" is ambiguous

/ascldap/users/jjellio/src/Trilinos/packages/teuchos/parser/test/unit_tests.cpp(433): error: "<unnamed>::any" is ambiguous

/ascldap/users/jjellio/src/Trilinos/packages/teuchos/parser/test/unit_tests.cpp(455): error: "<unnamed>::any" is ambiguous

/ascldap/users/jjellio/src/Trilinos/packages/teuchos/parser/test/unit_tests.cpp(479): error: "<unnamed>::any" is ambiguous

/ascldap/users/jjellio/src/Trilinos/packages/teuchos/parser/test/unit_tests.cpp(491): error: "<unnamed>::any" is ambiguous

7 errors detected in the compilation of "/tmp/tmpxft_00006861_00000000-7_unit_tests.cpp1.ii".
make[2]: *** [packages/teuchos/parser/test/CMakeFiles/TeuchosParser_Parser_UnitTests.dir/unit_tests.cpp.o] Error 2
make[1]: *** [packages/teuchos/parser/test/CMakeFiles/TeuchosParser_Parser_UnitTests.dir/all] Error 2
make: *** [all] Error 2
```

The error seems to occur in master or develop. I suppose this was not caught, because the GPU testing builds likely do not build all unit tests.

@trilinos/teuchos

**Comment by bartlettroscoe** [comment]
@ibaned,

FYI: Looks like this is an error in the new TeuchosParser package code.

**Comment by ibaned** [comment]
@bartlettroscoe Thanks! It looks like my request to join @trilinos/teuchos was approved, so I now get notified of issues where @trilinos/teuchos is mentioned. This seems easy to fix, I'll do this soon.

I suspect that the root cause is that CUDA defines the global symbol `::any`, and I believe it is a function which has nothing to do with C++17's `std::any`, so that is a forward-looking concern of CUDA+C++17 compatibility, but anyways I think if I fully namespace-qualify these symbols the errors should go away.

**Comment by ibaned** [comment]
Okay, this should be fixed by the merge of #1898. If TeuchosParser still fails to build in this configuration, please reopen this issue.
